wgengine/router/router_windows.go

```diff
@@ -171,75 +210,128 @@ func (ft *firewallTweaker) doAsyncSet() {
 
 	ft.mu.Lock()
 	for { // invariant: ft.mu must be locked when beginning this block
-		val := ft.want
-		if ft.known && strsEqual(ft.lastVal, val) {
+		val := ft.wantLocal
+		if ft.known && strsEqual(ft.lastLocal, val) && ft.wantKillswitch == ft.lastKillswitch {
 			ft.running = false
 			ft.logf("ending netsh goroutine")
 			ft.mu.Unlock()
 			return
 		}
-		needClear := !ft.known || len(ft.lastVal) > 0 || len(val) == 0
+		wantKillswitch := ft.wantKillswitch
+		needClear := !ft.known || len(ft.lastLocal) > 0 || len(val) == 0
 		needProcRule := !ft.didProcRule
 		ft.mu.Unlock()
 
-		if needClear {
-			ft.logf("clearing Tailscale-In firewall rules...")
-			// We ignore the error here, because netsh returns an error for
-			// deleting something that doesn't match.
-			// TODO(bradfitz): care? That'd involve querying it before/after to see
-			// whether it was necessary/worked. But the output format is localized,
-			// so can't rely on parsing English. Maybe need to use OLE, not netsh.exe?
-			d, _ := ft.runFirewall("delete", "rule", "name=Tailscale-In", "dir=in")
-			ft.logf("cleared Tailscale-In firewall rules in %v", d)
+		err := ft.doSet(val, wantKillswitch, needClear, needProcRule)
+		bo.BackOff(ctx, err)
+
+		ft.mu.Lock()
+		ft.lastLocal = val
+		ft.lastKillswitch = wantKillswitch
+		ft.known = (err == nil)
+	}
+}
+
+func (ft *firewallTweaker) doSet(local []string, killswitch bool, clear bool, procRule bool) error {
```

Comment:
"a, b, c bool"

also, worth documenting this signature a bit? lot of stuff going on.

Reply:
The bools aren't great, but doesn't seem worth making an options struct given that I'm planning to replace most of this code with WFP API things after 1.6.

Added more documentation. The main purpose of doSet is to be a func that can return an error for BackOff in doAsyncSet, rather than have to do subtle breaks and gotos to get the control flow right.

Reply:
I'm not saying write a struct. Just to merge multiple args of the same type:

```patch
- func (ft *firewallTweaker) doSet(local []string, killswitch bool, clear bool, procRule bool) error {
+ func (ft *firewallTweaker) doSet(local []string, killswitch, clear, procRule bool) error {
```